enterprise/functions/src/main/java/io/crate/window/LagFunction.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * This file is part of a module with proprietary Enterprise Features.
+ *
+ * Licensed to Crate.io Inc. ("Crate.io") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ *
+ * Unauthorized copying of this file, via any medium is strictly prohibited.
+ *
+ * To use this file, Crate.io must have given you permission to enable and
+ * use such Enterprise Features and you must have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  If you enable or use the Enterprise
+ * Features, you represent and warrant that you have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  Your use of the Enterprise Features
+ * if governed by the terms and conditions of your Enterprise or Subscription
+ * Agreement with Crate.io.
+ */
+
+package io.crate.window;
+
+import io.crate.data.Input;
+import io.crate.data.Row;
+import io.crate.data.RowN;
+import io.crate.execution.engine.collect.CollectExpression;
+import io.crate.execution.engine.window.WindowFrameState;
+import io.crate.execution.engine.window.WindowFunction;
+import io.crate.metadata.BaseFunctionResolver;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionImplementation;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.functions.params.FuncParams;
+import io.crate.metadata.functions.params.Param;
+import io.crate.module.EnterpriseFunctionsModule;
+import io.crate.types.DataType;
+
+import java.util.List;
+
+public class LagFunction implements WindowFunction {
+
+    private static final String NAME = "lag";
+
+    private final FunctionInfo info;
+
+    private LagFunction(FunctionInfo info) {
+        this.info = info;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @Override
+    public Object execute(int idxInPartition,
+                          WindowFrameState currentFrame,
+                          List<? extends CollectExpression<Row, ?>> expressions,
+                          Input... args) {
+
+        final int offset;
+        if (args.length > 1) {
+            offset = ((Number) args[1].value()).intValue();
+        } else {
+            offset = 1;
+        }
+
+        if (offset <= idxInPartition) {
```

Comment:
Could this check be removed and instead rely on `currentFrame.getRowInPartitionAtIndexOrNull` returning `null` - in which case the default is returned?

Reply:
not sure, couldn't really achieve that, coz once you have more than one partition , you cannot rely any longer on `getRowInPartitionAtIndexOrNull` to fetch the lag row, coz it will fetch it from the previous partition.

Reply:
I don't think this is the case as a WindowFrameState has the rows in only one partition

Reply:
Should `getRowInFrameAtIndexOrNull` be used instead? Is it okay for `lag` to return values that are outside of the current frame but within a partition?

Reply:
`lag` does not operate on a frame, but on a partition, so i'd say it is ok. it is not really possible to use `getRowInFrameAtIndexOrNull`, coz for cases when a window shrinks, lag rows won't be accessible. See

```
   public Object[] getRowInFrameAtIndexOrNull(int index) {
        if (index < lowerBound || index >= upperBoundExclusive) {
            return null;
        }
```

Reply:
@andreidan it does not seem to be true for all cases :/ For the [test case](https://github.com/crate/crate/pull/8880/files#diff-30efabebfd5962f651f4b8d63553957fR135) the `WindowFrameState` keeps all rows and partitions are defined/distinguished by the `partitionStart` index.

Reply:
Yep, that's true, sorry, it carries the rows for the entire window

Reply:
I got rid of the condition. now it is fine to use `getRowInPartitionAtIndexOrNull` to get a row within a partition. See the first commit in the PR that allows to do that.